project-metrics/metrics_service/metrics/presubmit_latency.py

```diff
@@ -1,55 +1,27 @@
-"""Presubmit Latency metric."""
+"""CircleCI Presubmit Latency."""
 
-import sqlalchemy
 from typing import Text
-
-from database import db
-from database import models
+from metrics.base import Metric
 from metrics import base
+from database import models
+from apis.circleci import CircleCiAPI
 
-
-class PresubmitLatencyMetric(base.Metric):
-  """A metric tracking the average duration of Travis PR builds."""
-
-  UNIT = 'seconds'
-
-  def _format_value(self, avg_seconds: float) -> Text:
-    return '%dm' % (avg_seconds // 60)
-
-  def _score_value(self, avg_seconds: float) -> models.MetricScore:
-    avg_minutes = avg_seconds / 60
-    if avg_minutes > 35:
-      return models.MetricScore.CRITICAL
-    elif avg_minutes > 25:
+class PresubmitLatency(Metric):
+  def _score_value(self, percentage: float) -> models.MetricScore:
+    if percentage < 0.6:
       return models.MetricScore.POOR
-    elif avg_minutes > 15:
+    elif percentage < 0.75:
       return models.MetricScore.MODERATE
-    elif avg_minutes > 10:
+    elif percentage < 0.9:
       return models.MetricScore.GOOD
     else:
       return models.MetricScore.EXCELLENT
 
-  def _compute_value(self) -> float:
-    """Computes the average duration of all completed builds.
-
-    Excludes builds that are newly created, pending, cancelled, or errored since
-    these either have no duration or are not representative of a real build.
-
-    Raises:
-      ValueError: if no builds exist.
-
-    Returns:
-      The percentage of passing builds.
-    """
-    session = db.Session()
-    avg_seconds = session.query(sqlalchemy.func.avg(
-        models.Build.duration)).filter(
-            models.Build.is_last_90_days(base_time=self.base_time)).scalar()
-    session.close()
+  def _compute_value(self) -> models.MetricResult:
+    workflow_stats = CircleCiAPI().get_workflow_stats()
```

Comment:
FWIW this means each metric will separately be requesting the same stats from the `CircleCiApi`. Unlikely we'd hit rate-limits like this, but something to be aware of. Also, this does not appear to limit the stats to a 90-day window using `self.base_time`. `self.base_time` was added per request from...Raghu? Will? idk, but supporting it allows for back-filling data by some amount of time. I don't know if the CircleCI API supports arbitrary time windows, though. This was also why we were scraping previous build data and computing the stats ourselves; it meant we could always reproduce/recompute the values.

Reply:
The function `get_workflow_stats` has a default argument of `last-90-days` so that much will work, however, circleci's api does not currently allow arbitrary time intervals for their insights api. I do not believe that arbirary intervals are a requirement anymore so, while it would be possible to manually compute these values by paginating their workflows api, I don't see that as being necessary. @rsimha